enhancements/local-storage/health-monitoring.md

```diff
@@ -0,0 +1,137 @@
+authors:
+  - "@rohantmp"
+reviewers:
+  - "@gnufied"
+  - "@jsafrane”
+  - "@sp98"
+  
+approvers:
+  - "@gnufied"
+  - "@jsafrane"
+creation-date: 2021-07-16
+last-updated: 2021-07-16
+status: implementable
+see-also:
+replaces:
+superseded-by:
+---
+
+
+
+# Local Device Health Monitoring
+
+
+## Release Signoff Checklist
+
+- [x] Enhancement is `implementable`
+- [x] Design details are appropriately documented from clear requirements
+- [x] Test plan is defined
+- [ ] Operational readiness criteria is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Summary
+
+This enhancement proposes that we export health metrics for each local device on each node.
+
+## Motivation
+
+Local device vendors expose health information about storage devices via APIs like [SMART](https://en.wikipedia.org/wiki/S.M.A.R.T.). It very useful to know when a storage is about to fail, to differentiate between storage hardware related errors vs purely software ones, etc.
+
+### Goals
+
+- Expose health information about local storage devices in a way that can generate alerts
+
+### Non-Goals
+
+- Align with sig-storage [health monitoring for persistent volumes](https://kubernetes.io/blog/2021/04/16/volume-health-monitoring-alpha-update/)
+  - justification: it is also useful to monitor system drives and CSI only deals with persitent volumes, and migrating to CSI is an extremely long term effort.
+
+## Proposal
+
+[Local Storage Operator](https://github.com/openshift/local-storage-operator) currently deploys multiple daemonsets and enumerates local-storage. We could export per-disk metrics from the same daemonset.
+
+
+### User Stories
+
+- As an admin, I would like to be alerted when a storage device is unhealthy and be linked to remediation steps.
+- As an admin, I would like if components (such as OpenShift Data Foundation) could access storage device health and failover faster.
+
+
+### Implementation Details/Notes/Constraints
+
+`LocalVolumeDisocveryResults` already enumerate and refresh information about the disk in a per-node `LocalVolumeDiscoveryResult`, and a [PR](https://github.com/openshift/local-storage-operator/pull/249) is in progress to add metrics to each of the daemonsets that update the CR.
+
+`LocalVolumeDiscovery` can be configured via `nodeSelector` and `toleration` to match any set of nodes or all nodes. We could have the metrics be for all nodes or just for the ones that LocalVolumeDiscovery is configured to watch. Whether this should be controlled by the same (singleton) CR is an [open question](#Open-Questions    )
+
+We will be collecting metrics via [libstoragemgmt](https://github.com/libstorage/libstoragemgmt) and calling it via [libstoragemgmt-golang](https://github.com/libstorage/libstoragemgmt-golang/).
+If a block device is a partion or logical volume, we will only expose health for the parent device.
```

Comment:
In case of LVM it can be array of devices. Whose health will be reported?

Reply:
I guess this was for when we were considering per-PV metrics. LVM volumes count as a logical volume. We wouldn't expose metrics for them, just their parents.